File: examples/low_rank/low_rank.py
hlp = """
    Evaluation of low-rank kernel approximation methods
    with ridge regression on standard datasets from KEEL.
"""

import os
import csv
import sys
import itertools as it
import scipy.stats as st
import time
import argparse

# Low-rank approximation methods
from mklaren.regression.ridge import RidgeLowRank
from mklaren.regression.ridge import RidgeMKL
from mklaren.regression.fitc import FITC
from mklaren.projection.rff import RFF
from mklaren.mkl.mklaren import Mklaren

# Kernels
from mklaren.kernel.kernel import exponential_kernel, kernel_sum
from mklaren.kernel.kinterface import Kinterface

# Datasets
from datasets.keel import load_keel, KEEL_DATASETS

# Utils
from numpy import var, mean, logspace
from random import shuffle, seed
from math import ceil

# Comparable parameters with effective methods
N               = 3000                            # Max number of data points
rank_range      = range(2, 100)                   # Rang of tested ranks
p_range         = (10,)                           # Fixed number of kernels
lbd_range       = [0] + list(logspace(-5, 1, 7))  # Regularization parameter
delta           = 10                              # Number of look-ahead columns (CSI and mklaren)
cv_iter         = 5                               # Cross-validation iterations
training_size   = 0.6                             # Training set fraction
validation_size = 0.2                             # Validation set (for fitting hyperparameters)
test_size       = 0.2                             # Test set (for reporting scores)


# Method classes and fixed hyperparameters
methods = {
    "CSI" :        (RidgeLowRank, {"method": "csi", "method_init_args": {"delta": delta}}),
    "ICD" :        (RidgeLowRank, {"method": "icd"}),
    "Nystrom":     (RidgeLowRank, {"method": "nystrom"}),
    "CSI*" :       (RidgeLowRank, {"method": "csi", "method_init_args": {"delta": delta}}),
    "ICD*" :       (RidgeLowRank, {"method": "icd"}),
    "Nystrom*":    (RidgeLowRank, {"method": "nystrom"}),
    "Mklaren":     (Mklaren,      {"delta": delta}),
    "RFF":         (RFF,          {"delta": delta}),
    "FITC":        (FITC,         {}),
    "uniform":     (RidgeMKL,     {"method": "uniform"}),
    "L2KRR":       (RidgeMKL,     {"method": "l2krr"}),
}


def process(dataset, outdir):
    """
    Run experiments with specified parameters.
    :param dataset: Dataset key.
    :param outdir: Output directory.
    :return:
    """

    # Fixed output
    # Create output directory
    if not os.path.exists(outdir): os.makedirs(outdir)
    fname = os.path.join(outdir, "%s.csv" % dataset)

    # Output
    header = ["dataset", "n", "method", "rank", "erank", "iteration", "lambda",
              "gmin", "gmax", "p", "evar_tr", "evar", "time",
              "RMSE_tr", "RMSE_va", "RMSE"]
    fp = open(fname, "w", buffering=0)
    writer = csv.DictWriter(fp, fieldnames=header)
    writer.writeheader()

    # Load data
    data = load_keel(n=N, name=dataset)

    # Load data and normalize columns
    X = st.zscore(data["data"], axis=0)
    y = st.zscore(data["target"])
    n = len(X)

    # Perform model cross-validation with internal parameter selection
    for cv, p in it.product(range(cv_iter), p_range):
        gam_range = logspace(-6, 3, p, base=2)

        # Split into training, validation, test sets.
        seed(cv)
        inxs = range(len(X))
        shuffle(inxs)
        n1 = int(training_size * n)
        n2 = int(validation_size * n)
        tr, tval, te = inxs[:n1], inxs[n1:n1+n2], inxs[n1+n2:]
        X_tr, X_val, X_te = X[tr], X[tval], X[te]
        y_tr, y_val, y_te = y[tr], y[tval], y[te]

        # Store kernel interfaces for training data
        Ks_full = [Kinterface(data=X,
                         kernel=exponential_kernel,
                         kernel_args={"gamma": gam}) for gam in gam_range]
        Ks = [Kinterface(data=X_tr,
                         kernel=exponential_kernel,
                         kernel_args={"gamma": gam}) for gam in gam_range]
        Ksum = Kinterface(data=X_tr,
                          kernel=kernel_sum,
                          kernel_args={"kernels": [exponential_kernel] * len(gam_range),
                                       "kernels_args": [{"gamma": gam} for gam in gam_range]})

        for mname, (Mclass, kwargs) in methods.items():

            # Fit hyperparameters on the validation set using the current rank
            for rank in rank_range:
                for lbd in lbd_range:
                    times = []
                    try:
                        t_train = time.time()
                        if mname == "Mklaren":
                            effective_rank = rank
                            model = Mclass(lbd=lbd, rank=rank, **kwargs)
                            model.fit(Ks, y_tr)
                            yptr    = model.predict([X_tr] * len(Ks)).ravel()
                            ypva    = model.predict([X_val] * len(Ks)).ravel()
                            ypte    = model.predict([X_te] * len(Ks)).ravel()
                        elif mname == "RFF":
                            effective_rank = rank
                            model = Mclass(rank=rank, lbd=lbd,
                                           gamma_range=gam_range, **kwargs)
                            model.fit(X_tr, y_tr)
                            yptr = model.predict(X_tr).ravel()
                            ypva = model.predict(X_val).ravel()
                            ypte = model.predict(X_te).ravel()
                        elif mname == "FITC":
                            effective_rank = rank
                            model = Mclass(rank=rank, **kwargs)
                            model.fit(Ks, y_tr)
                            yptr = model.predict([X_tr] * len(Ks)).ravel()
                            ypva = model.predict([X_val]* len(Ks)).ravel()
                            ypte = model.predict([X_te] * len(Ks)).ravel()
                        elif mname in ("uniform", "L2KRR"):
                            effective_rank = rank
                            model = Mclass(lbd=lbd, **kwargs)
                            model.fit(Ks_full, y, holdout=te+tval)
                            yptr = model.predict(tr).ravel()
                            ypva = model.predict(tval).ravel()
                            ypte = model.predict(te).ravel()
                        elif mname in ("CSI*", "ICD*", "Nystrom*"):   # Separate approximations
                            effective_rank = int(max(1, ceil(1.0 * rank / p)))
                            model = Mclass(lbd=lbd, rank=effective_rank, **kwargs)
                            model.fit(Ks, y_tr)
                            yptr = model.predict([X_tr] * len(Ks)).ravel()
                            ypva = model.predict([X_val] * len(Ks)).ravel()
                            ypte = model.predict([X_te] * len(Ks)).ravel()
                        else:   # Other low-rank approximations; Mklaren2
                            effective_rank = rank
                            model = Mclass(lbd=lbd, rank=rank, **kwargs)
                            model.fit([Ksum], y_tr)
                            yptr = model.predict([X_tr]).ravel()
                            ypva = model.predict([X_val]).ravel()
                            ypte = model.predict([X_te]).ravel()
                        t_train = time.time() - t_train
                        times.append(t_train)
                    except Exception as e:
                        sys.stderr.write("Method: %s rank: %d iter: %d error: %s \n" % (mname, rank, cv, str(e)))
                        continue

                    # Compute errors
                    score_tr = var(y_tr - yptr) ** 0.5
                    score_va = var(y_val - ypva) ** 0.5
                    score_te = var(y_te - ypte) ** 0.5

                    # Explained variance
                    evar_tr = (var(y_tr) - var(y_tr - yptr)) / var(y_tr)
                    evar_te = (var(y_te) - var(y_te - ypte)) / var(y_te)

                    # Write to output
                    row = {"dataset": dataset, "method": mname, "rank": rank, "n": n,
                           "iteration": cv, "lambda": lbd, "time": mean(times),
                           "evar": evar_te, "evar_tr": evar_tr, "erank": effective_rank,
                           "RMSE": score_te, "RMSE_va": score_va, "RMSE_tr": score_tr,
                           "gmin": min(gam_range), "gmax": max(gam_range), "p": len(gam_range)}
                    writer.writerow(row)

                    # Break for FITC / no lambda
                    if mname == "FITC": break

                # Break for FITC / no rank
                if mname in ("uniform", "L2KRR"): break


if __name__ == "__main__":
    # Input arguments
    parser = argparse.ArgumentParser(description=hlp)
    parser.add_argument("dataset", help="Dataset. One of {%s}." % ", ".join(KEEL_DATASETS))
    parser.add_argument("output",  help="Output directory.")
    args = parser.parse_args()

    # Output directory
    data_set = args.dataset
    out_dir = args.output
    assert data_set in KEEL_DATASETS
    process(data_set, out_dir)

File: examples/low_rank/low_rank.R
hlp = "Post processing of the results on low-rank approximations
       (KEEL datasets)."

require(optparse)
require(scmamp)
require(xtable)

# Parse input arguments
option_list = list(
  make_option(c("-i", "--input"), type="character", help="Results file (.csv)"),
  make_option(c("-o", "--output"), type="character", help="Output directory")
);
opt_parser = OptionParser(option_list=option_list, description=hlp);
opt = parse_args(opt_parser);
in_file = opt$input
out_dir = opt$output

# Target methods
target = "L2KRR" 

# Low rank approximations
lowranks = c("Mklaren", "CSI", "ICD", "Nystrom", "CSI*", "ICD*",  "Nystrom*", "RFF", "RFF-NS", "SPGP")
matrices = c("Mklaren", "CSI", "ICD", "Nystrom", "CSI*", "ICD*",  "Nystrom*")
fullranks = c("L2KRR", "uniform", "Mklaren2")

# Load
alldata = read.csv(in_file, header=TRUE, stringsAsFactors = FALSE)

# Aggregate results 
meths = sort(setdiff(unique(alldata$method), fullranks))
datasets = unique(alldata$dataset)
R = matrix(NA, nrow=length(datasets), ncol=length(meths))
row.names(R) = datasets
colnames(R) = meths

# Store times
Time = matrix(NA, nrow=length(datasets), ncol=length(meths))
row.names(Time) = datasets
colnames(Time) = meths

# Select lambda via cross-validation
agg = aggregate(alldata$RMSE_va, by=list(dataset=alldata$dataset,
                                         method=alldata$method,
                                         rank=alldata$rank,
                                         iteration=alldata$iteration,
                                         p=alldata$p), min)
row.names(agg) <- sprintf("%s.%s.%d.%s.%d", agg$dataset, agg$method, agg$rank, agg$iteration, agg$p)
inxs = sprintf("%s.%s.%d.%s.%d", alldata$dataset, alldata$method, 
               alldata$rank, alldata$iteration, alldata$p)
alldata$best = alldata$RMSE_va == agg[inxs, "x"]
data = alldata[alldata$best,]

# Aggregate mean and sd per rank
agg.m = aggregate(data[, "RMSE"], by=list(dataset=data[,"dataset"], method=data[, "method"],
                    eff.rank=data[, "eff.rank"]), mean)
agg.s = aggregate(data[, "RMSE"], by=list(dataset=data[,"dataset"], method=data[, "method"],
                    eff.rank=data[, "eff.rank"]), sd)

# Aggregate mean and sd per rank
inxs = agg.m$method == target
t = agg.m[inxs, "x"] + agg.s[inxs, "x"]  
names(t) <- agg.m[inxs, "dataset"]

# Select valid entries
filt = agg.m$x < t[agg.m$dataset] & agg.m$method %in% meths
agg.valid = agg.m[filt,]
agg.min = aggregate(agg.valid$eff.rank, by=list(dataset=agg.valid$dataset, method=agg.valid$method), min) 

# Extract minimal rank and corresponding time
for (i in 1:nrow(agg.min)){
  R[agg.min[i, "dataset"], agg.min[i, "method"]] = agg.min[i, "x"]
  df = subset(data, dataset == agg.min[i, "dataset"] & 
                       rank == agg.min[i, "x"] & 
                     method == agg.min[i, "method"])
  t = mean(abs(df$time))
  Time[agg.min[i, "dataset"], agg.min[i, "method"]] = t
}

# Rank ; eliminate rows with all inifinite rank
excluded = rowSums(is.na(R[,matrices])) >= length(matrices)
message("Excluded datasets:")
print(row.names(R)[excluded])

zout = rowSums(is.na(R[,matrices])) < length(matrices)
R[is.na(R)] = Inf
R = R[zout,]
rnk = apply(R, MARGIN = 1, FUN=rank)
rm = sort(rowMeans(rnk))
message("Mean ranks:")
print(rm)


# Friedman test on ranks
message("Friedman test on ranks:")
friedman.test(R)

# Compare Mklaren and CSI
message("Wilcoxon test between Mklaren and CSI:")
print(wilcox.test(R[,c("Mklaren")], R[,c("CSI")], paired=TRUE, alternative="less"))

# Store CD
fname = file.path(out_dir, "rank_CD.pdf")
pdf(fname)
plotCD(-R, alpha = 0.05)
dev.off()
message(sprintf("Written %s", fname))

# Store CD
fname = file.path(out_dir, "rank_CD.eps")
postscript(fname)
plotCD(-R, alpha = 0.05)
dev.off()
message(sprintf("Written %s", fname))

# Dataset names to n 
dn = data[,c("dataset", "n")]
dn = dn[!duplicated(dn) & !is.na(dn$dataset),]
row.names(dn) = dn$dataset
n = dn$n
names(n) = row.names(dn)
n = n[row.names(R)]

# Store a LaTeX table
Rn = cbind(n, R[,lowranks])
Rn = Rn[order(Rn[,"n"]),]
Rt = matrix("", nrow=nrow(Rn), ncol=ncol(Rn))
colnames(Rt) = colnames(Rn)
row.names(Rt) = row.names(Rn)
for (i in 1:nrow(Rn)){
  minn = names(Rn[i,matrices])[Rn[i,matrices] == min(Rn[i,matrices])]
  Rt[i,] = Rn[i,]
  Rt[i,minn] = sprintf("\\textbf{%s}", Rn[i, minn]) 
}

# Write to output
fname = file.path(out_dir, "ranks.tex")
tab = xtable(Rt)
align(tab) = c("r", "l", "|", "|", "l", "l", "l", "l", "|", 
               "l", "l", "l", "|", "l", "l", "|", "l")
sink(fname)
print(tab, sanitize.text.function = identity)
sink()